File: go/resolution_conformance_test.go
package skg

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type resolutionContract struct {
	ContractVersion int                      `json:"contract_version"`
	LanguageVersion string                   `json:"language_version"`
	Cases           []resolutionContractCase `json:"cases"`
}

type resolutionContractCase struct {
	Name   string `json:"name"`
	Entry  string `json:"entry"`
	Root   string `json:"root"`
	Limits struct {
		Bytes     int64 `json:"bytes"`
		Files     int   `json:"files"`
		Nodes     int   `json:"nodes"`
		MergeWork int64 `json:"merge_work"`
	} `json:"limits"`
	Files []struct {
		Path   string  `json:"path"`
		Source *string `json:"source"`
	} `json:"files"`
	ExpectedFormatted *string    `json:"expected_formatted"`
	ExpectedCode      *ErrorCode `json:"expected_code"`
}

func TestResolutionConformance(t *testing.T) {
	const path = "../testdata/resolution/cases.json"
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	validateResolutionContractJSON(t, data)

	var suite resolutionContract
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&suite); err != nil {
		t.Fatal(err)
	}
	if err := decoder.Decode(new(any)); err != io.EOF {
		t.Fatal("trailing JSON value")
	}
	if suite.ContractVersion != 1 || suite.LanguageVersion != "1.0" || len(suite.Cases) == 0 {
		t.Fatalf("unsupported or empty resolution contract: version=%d language=%q cases=%d", suite.ContractVersion, suite.LanguageVersion, len(suite.Cases))
	}

	codes := loadErrorCodes(t)
	names := make(map[string]bool, len(suite.Cases))
	for _, fixture := range suite.Cases {
		fixture := fixture
		if fixture.Name == "" || names[fixture.Name] {
			t.Fatalf("empty or duplicate resolution case name %q", fixture.Name)
		}
		if fixture.ExpectedCode != nil && !codes[string(*fixture.ExpectedCode)] {
			t.Fatalf("resolution case %q names unregistered error code %q", fixture.Name, *fixture.ExpectedCode)
		}
		names[fixture.Name] = true
		t.Run(fixture.Name, func(t *testing.T) { runResolutionContractCase(t, fixture) })
	}
	t.Logf("resolution contract v%d: %d cases, none skipped", suite.ContractVersion, len(suite.Cases))
}

func validateResolutionContractJSON(t *testing.T, data []byte) {
	t.Helper()
	root := nativeFixtureObject(t, data,
		[]string{"contract_version", "language_version", "cases"},
		[]string{"contract_version", "language_version", "cases"})
	var cases []json.RawMessage
	if err := json.Unmarshal(root["cases"], &cases); err != nil {
		t.Fatal(err)
	}
	for _, raw := range cases {
		obj := nativeFixtureObject(t, raw,
			[]string{"name", "entry", "root", "limits", "files", "expected_formatted", "expected_code"},
			[]string{"name", "entry", "limits", "files"})
		_, formatted := obj["expected_formatted"]
		_, code := obj["expected_code"]
		if formatted == code {
			t.Fatal("resolution case needs exactly one of expected_formatted and expected_code")
		}
		nativeFixtureObject(t, obj["limits"],
			[]string{"bytes", "files", "nodes", "merge_work"},
			[]string{"bytes", "files", "nodes", "merge_work"})
		var files []json.RawMessage
		if err := json.Unmarshal(obj["files"], &files); err != nil {
			t.Fatal(err)
		}
		for _, file := range files {
			nativeFixtureObject(t, file, []string{"path", "source"}, []string{"path", "source"})
		}
	}
}

func runResolutionContractCase(t *testing.T, fixture resolutionContractCase) {
	t.Helper()
	if fixture.Entry == "" || len(fixture.Files) == 0 || fixture.Limits.Bytes <= 0 || fixture.Limits.Files <= 0 || fixture.Limits.Nodes <= 0 || fixture.Limits.MergeWork <= 0 {
		t.Fatal("entry, files, and positive explicit limits are required")
	}
	base := t.TempDir()
	seen := make(map[string]bool, len(fixture.Files))
	for _, file := range fixture.Files {
		path := safeContractPath(t, base, file.Path)
		if file.Source == nil || seen[path] {
			t.Fatalf("file %q has null source or duplicate path", file.Path)
		}
		seen[path] = true
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			t.Fatal(err)
		}
		if err := os.WriteFile(path, []byte(*file.Source), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	options := ResolveOptions{
		MaxBytes: fixture.Limits.Bytes, MaxFiles: fixture.Limits.Files,
		MaxNodes: fixture.Limits.Nodes, MaxMergeWork: fixture.Limits.MergeWork,
	}
	if fixture.Root != "" {
		options.Root = safeContractPath(t, base, fixture.Root)
	}
	file, err := ParseFileWithOptions(safeContractPath(t, base, fixture.Entry), options)
	if fixture.ExpectedCode != nil {
		if err == nil {
			t.Fatalf("want %s, got success", *fixture.ExpectedCode)
		}
		var parseErr *ParseError
		if !errors.As(err, &parseErr) {
			t.Fatalf("want *ParseError with %s, got %T: %v", *fixture.ExpectedCode, err, err)
		}
		if parseErr.Diag.Code != *fixture.ExpectedCode {
			t.Fatalf("want %s, got %s: %v", *fixture.ExpectedCode, parseErr.Diag.Code, err)
		}
		return
	}
	if err != nil {
		t.Fatal(err)
	}
	if got := string(Emit(file)); got != *fixture.ExpectedFormatted {
		t.Fatalf("resolved output mismatch\nwant:\n%s\ngot:\n%s", *fixture.ExpectedFormatted, got)
	}
}

func safeContractPath(t *testing.T, base, slashPath string) string {
	t.Helper()
	path := filepath.Clean(filepath.FromSlash(slashPath))
	if slashPath == "" || filepath.IsAbs(path) || path == ".." || strings.HasPrefix(path, ".."+string(filepath.Separator)) {
		t.Fatalf("unsafe resolution contract path %q", slashPath)
	}
	return filepath.Join(base, path)
}

File: zig/resolution_conformance_test.zig
//! Shared import-resolution, containment, and aggregate-budget contract.
const std = @import("std");
const testing = std.testing;
const skg = @import("root.zig");

const Limits = struct {
    bytes: usize,
    files: usize,
    nodes: usize,
    merge_work: usize,
};

const ContractFile = struct {
    path: []const u8,
    source: []const u8,
};

const Fixture = struct {
    name: []const u8,
    entry: []const u8,
    root: ?[]const u8 = null,
    limits: Limits,
    files: []ContractFile,
    expected_formatted: ?[]const u8 = null,
    expected_code: ?skg.ast.ErrorCode = null,
};

const Suite = struct {
    contract_version: u32,
    language_version: []const u8,
    cases: []Fixture,
};

test "shared resolution conformance" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const data = try std.fs.cwd().readFileAlloc(allocator, "testdata/resolution/cases.json", 1024 * 1024);
    const raw = try std.json.parseFromSlice(std.json.Value, allocator, data, .{});
    try validateContractShape(raw.value);
    const suite = (try std.json.parseFromValue(Suite, allocator, raw.value, .{})).value;

    try testing.expectEqual(@as(u32, 1), suite.contract_version);
    try testing.expectEqualStrings("1.0", suite.language_version);
    try testing.expect(suite.cases.len > 0);
    var names = std.StringHashMap(void).init(allocator);
    for (suite.cases) |fixture| {
        try testing.expect(fixture.name.len > 0);
        try testing.expect(!names.contains(fixture.name));
        try names.put(fixture.name, {});
        runFixture(fixture) catch |err| {
            std.debug.print("resolution fixture {s} failed: {}\n", .{ fixture.name, err });
            return err;
        };
    }
    std.debug.print("resolution contract v{d}: {d} cases, none skipped\n", .{ suite.contract_version, suite.cases.len });
}

fn validateContractShape(value: std.json.Value) !void {
    const root = try object(value);
    try fields(root, &.{ "contract_version", "language_version", "cases" }, &.{ "contract_version", "language_version", "cases" });
    const cases = switch (root.get("cases").?) {
        .array => |array| array.items,
        else => return error.BadResolutionContract,
    };
    for (cases) |case_value| {
        const case = try object(case_value);
        try fields(
            case,
            &.{ "name", "entry", "root", "limits", "files", "expected_formatted", "expected_code" },
            &.{ "name", "entry", "limits", "files" },
        );
        if (case.contains("expected_formatted") == case.contains("expected_code")) return error.BadResolutionContract;
        if (case.get("expected_formatted")) |expected| if (expected == .null) return error.BadResolutionContract;
        if (case.get("expected_code")) |expected| if (expected == .null) return error.BadResolutionContract;
        const limits = try object(case.get("limits").?);
        try fields(limits, &.{ "bytes", "files", "nodes", "merge_work" }, &.{ "bytes", "files", "nodes", "merge_work" });
        const files_value = case.get("files").?;
        const file_values = switch (files_value) {
            .array => |array| array.items,
            else => return error.BadResolutionContract,
        };
        for (file_values) |file_value| {
            const file = try object(file_value);
            try fields(file, &.{ "path", "source" }, &.{ "path", "source" });
        }
    }
}

fn object(value: std.json.Value) !std.json.ObjectMap {
    return switch (value) {
        .object => |map| map,
        else => error.BadResolutionContract,
    };
}

fn fields(map: std.json.ObjectMap, allowed: []const []const u8, required: []const []const u8) !void {
    var it = map.iterator();
    while (it.next()) |entry| {
        var known = false;
        for (allowed) |name| known = known or std.mem.eql(u8, name, entry.key_ptr.*);
        if (!known) return error.BadResolutionContract;
    }
    for (required) |name| if (!map.contains(name)) return error.BadResolutionContract;
}

fn runFixture(fixture: Fixture) !void {
    if (fixture.entry.len == 0 or fixture.files.len == 0 or fixture.limits.bytes == 0 or fixture.limits.files == 0 or fixture.limits.nodes == 0 or fixture.limits.merge_work == 0) {
        return error.BadResolutionContract;
    }
    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();
    var paths = std.StringHashMap(void).init(testing.allocator);
    defer paths.deinit();
    for (fixture.files) |file| {
        try validateRelativePath(file.path);
        if (paths.contains(file.path)) return error.DuplicateContractPath;
        try paths.put(file.path, {});
        if (std.fs.path.dirname(file.path)) |dir| try tmp.dir.makePath(dir);
        try tmp.dir.writeFile(.{ .sub_path = file.path, .data = file.source });
    }

    try validateRelativePath(fixture.entry);
    const entry = try tmp.dir.realpathAlloc(testing.allocator, fixture.entry);
    defer testing.allocator.free(entry);
    var options = skg.ResolveOptions{
        .max_bytes = fixture.limits.bytes,
        .max_files = fixture.limits.files,
        .max_nodes = fixture.limits.nodes,
        .max_merge_work = fixture.limits.merge_work,
    };
    var root_path: ?[]u8 = null;
    defer if (root_path) |path| testing.allocator.free(path);
    if (fixture.root) |root| {
        try validateRelativePath(root);
        root_path = try tmp.dir.realpathAlloc(testing.allocator, root);
        options.root = root_path.?;
    }

    var result = skg.parseWithOptions(testing.allocator, entry, options);
    defer result.deinit();
    if (fixture.expected_code) |code| {
        try testing.expect(result.file == null);
        const diagnostic = result.diagnostic orelse return error.MissingDiagnostic;
        try testing.expectEqual(code, diagnostic.code);
        return;
    }
    const file = result.file orelse {
        if (result.diagnostic) |diagnostic| {
            std.debug.print("unexpected {s}: {s}\n", .{ @tagName(diagnostic.code), diagnostic.message });
        }
        return error.UnexpectedResolutionFailure;
    };
    const emitted = try skg.emit.emitFile(testing.allocator, file);
    defer testing.allocator.free(emitted);
    try testing.expectEqualStrings(fixture.expected_formatted.?, emitted);
}

fn validateRelativePath(path: []const u8) !void {
    if (path.len == 0 or std.fs.path.isAbsolute(path) or std.mem.eql(u8, path, "..") or
        std.mem.startsWith(u8, path, "../") or std.mem.startsWith(u8, path, "..\\"))
    {
        return error.UnsafeContractPath;
    }
}
